### PR TITLE
Add Silhouette Score and Pairwise Distance

### PR DESCRIPTION
**Pull request by khizirsiddiqui**
Silhouette Score is a widely used metric for clustering algorithms, I thought it be nice to have it here.
 - Add Pairwise Distance (and test)
 - Add Silhouette Score (and test)
 - Add test for AssertSizes (already implemented, added test)

### CODE REVIEW

**Comment by birm** [comment]
Hi, this looks promising. Do you mind pulling in master?

**Comment by khizirsiddiqui** [comment]
@birm I did the rebase. Could you help with the documentation error? Is it because I am using a single equation inside `eqarray`?

File: src/mlpack/core/cv/metrics/silhouette_score.hpp [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+  static arma::rowvec SamplesScore(const DataType& X,
+                                   const arma::Row<size_t>& labels,
+                                   const Metric& metric);
+
+  /**
+   * Find mean distance of element from a given cluster.
+   *
+   * @param distances colvec containing distances from other elements.
+   * @param labels Labels assigned to data by clustering.
+   * @param label label of the target cluster.
```
Comment (birm) [inline src/mlpack/core/cv/metrics/silhouette_score.hpp:86]: Add doc for sameCluster

Comment (khizirsiddiqui) [reply src/mlpack/core/cv/metrics/silhouette_score.hpp:86]: :+1:

**Comment by birm** [comment]
I think this also needs a history.md line.

File: HISTORY.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
-### mlpack ?.?.?
+### mlpack 3.3.3
```
Comment (birm) [inline HISTORY.md:1]:
```suggestion
### mlpack ?.?.?
```
We note like that since it's not released, and it may, for example be 3.4.0; we don't know yet.

Comment (khizirsiddiqui) [reply HISTORY.md:1]: Oh, never knew this, thanks!

**Comment by birm** [comment]
@mlpack-jenkins test this please

**Comment by birm** [comment]
Oh, that's nice. Never done that before. :)

File: src/mlpack/core/cv/metrics/silhouette_score_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+  return arma::mean(SamplesScore(X, labels, metric));
+}
+
+template<typename DataType>
+arma::rowvec SilhouetteScore::SamplesScore(const DataType& distances,
+                                           const arma::Row<size_t>& labels)
+{
+  AssertSizes(distances, labels, "SilhouetteScore::SamplesScore()");
+
+  // Stores the silhouette scores of individual samples
```
Comment (birm) [inline src/mlpack/core/cv/metrics/silhouette_score_impl.hpp:35]:
```suggestion
  // Stores the silhouette scores of individual samples.
```

**Pull request merged**